chalice/deploy/packager.py

```diff
@@ -0,0 +1,213 @@
+import hashlib
+import inspect
+import os
+import shutil
+import subprocess
+import sys
+import zipfile
+
+import virtualenv
+from typing import Any, Optional  # noqa
+
+import chalice
+from chalice import compat, app
+
+
+class LambdaDeploymentPackager(object):
+    _CHALICE_LIB_DIR = 'chalicelib'
+    _VENDOR_DIR = 'vendor'
+
+    def _create_virtualenv(self, venv_dir):
+        # type: (str) -> None
+        # The original implementation used Popen(['virtualenv', ...])
+        # However, it's hard to make assumptions about how a users
+        # PATH is set up.  This could result in using old versions
+        # of virtualenv that give confusing error messages.
+        # To fix this issue, we're calling directly into the
+        # virtualenv package.  The main() method doesn't accept
+        # args, so we need to patch out sys.argv with the venv
+        # dir.  The original sys.argv is replaced on exit.
+        original = sys.argv
+        sys.argv = ['', venv_dir, '--quiet']
+        try:
+            virtualenv.main()
+        finally:
+            sys.argv = original
+
+    def create_deployment_package(self, project_dir, package_filename=None):
+        # type: (str, Optional[str]) -> str
+        print "Creating deployment package."
+        # pip install -t doesn't work out of the box with homebrew and
+        # python, so we're using virtualenvs instead which works in
+        # more cases.
+        venv_dir = os.path.join(project_dir, '.chalice', 'venv')
+        self._create_virtualenv(venv_dir)
+        pip_exe = compat.pip_script_in_venv(venv_dir)
+        assert os.path.isfile(pip_exe)
+        # Next install any requirements specified by the app.
+        requirements_file = os.path.join(project_dir, 'requirements.txt')
+        deployment_package_filename = self.deployment_package_filename(
+            project_dir)
+        if package_filename is None:
+            package_filename = deployment_package_filename
+        if self._has_at_least_one_package(requirements_file) and not \
+                os.path.isfile(package_filename):
+            p = subprocess.Popen([pip_exe, 'install', '-r', requirements_file],
+                                 stdout=subprocess.PIPE)
+            p.communicate()
+        deps_dir = compat.site_packages_dir_in_venv(venv_dir)
+        assert os.path.isdir(deps_dir)
+        # Now we need to create a zip file and add in the site-packages
+        # dir first, followed by the app_dir contents next.
+        dirname = os.path.dirname(os.path.abspath(package_filename))
+        if not os.path.isdir(dirname):
+            os.makedirs(dirname)
+        with zipfile.ZipFile(package_filename, 'w',
+                             compression=zipfile.ZIP_DEFLATED) as z:
+            self._add_py_deps(z, deps_dir)
```

Comment:
I think this would definitely be better.  That work is tracking in #249.

I should have clarified in the diff, but this `packager.py` code is all pre-existing, but is showing up as new code because I restructured some of the files.  I definitely want to circle back on this, a lot of this is old code that could use some rework.